src/mol-repr/volume/dot.ts

```diff
@@ -0,0 +1,274 @@
+    return props.tryUseImpostor && webgl && webgl.extensions.fragDepth && webgl.extensions.textureFloat
+        ? VolumeSphereImpostorVisual(materialId)
+        : VolumeSphereMeshVisual(materialId);
+}
+
+export function VolumeSphereImpostorVisual(materialId: number): VolumeVisual<VolumeSphereParams> {
+    return VolumeVisual<Spheres, VolumeSphereParams>({
+        defaultProps: PD.getDefaultValues(VolumeSphereParams),
+        createGeometry: createVolumeSphereImpostor,
+        createLocationIterator: createVolumeCellLocationIterator,
+        getLoci: getDotLoci,
+        eachLocation: eachDot,
+        setUpdateState: (state: VisualUpdateState, volume: Volume, newProps: PD.Values<VolumeSphereParams>, currentProps: PD.Values<VolumeSphereParams>) => {
+            state.createGeometry = (
+                !Volume.IsoValue.areSame(newProps.isoValue, currentProps.isoValue, volume.grid.stats)
+            );
+        },
+        geometryUtils: Spheres.Utils,
+        mustRecreate: (volumekey: VolumeKey, props: PD.Values<VolumeSphereParams>, webgl?: WebGLContext) => {
+            return !props.tryUseImpostor || !webgl;
+        }
+    }, materialId);
+}
+
+export function VolumeSphereMeshVisual(materialId: number): VolumeVisual<VolumeSphereParams> {
+    return VolumeVisual<Mesh, VolumeSphereParams>({
+        defaultProps: PD.getDefaultValues(VolumeSphereParams),
+        createGeometry: createVolumeSphereMesh,
+        createLocationIterator: createVolumeCellLocationIterator,
+        getLoci: getDotLoci,
+        eachLocation: eachDot,
+        setUpdateState: (state: VisualUpdateState, volume: Volume, newProps: PD.Values<VolumeSphereParams>, currentProps: PD.Values<VolumeSphereParams>) => {
+            state.createGeometry = (
+                !Volume.IsoValue.areSame(newProps.isoValue, currentProps.isoValue, volume.grid.stats) ||
+                newProps.sizeFactor !== currentProps.sizeFactor ||
+                newProps.detail !== currentProps.detail
+            );
+        },
+        geometryUtils: Mesh.Utils,
+        mustRecreate: (volumekey: VolumeKey, props: PD.Values<VolumeSphereParams>, webgl?: WebGLContext) => {
+            return props.tryUseImpostor && !!webgl;
+        }
+    }, materialId);
+}
+
+
+export function createVolumeSphereImpostor(ctx: VisualContext, volume: Volume, key: number, theme: Theme, props: VolumeSphereProps, spheres?: Spheres): Spheres {
+    const { cells: { space, data }, stats } = volume.grid;
+    const gridToCartn = Grid.getGridToCartesianTransform(volume.grid);
+    const isoVal = Volume.IsoValue.toAbsolute(props.isoValue, stats).absoluteValue;
+
+    const p = Vec3();
+    const [xn, yn, zn] = space.dimensions;
+
+    const count = Math.ceil((xn * yn * zn) / 10);
+    const builder = SpheresBuilder.create(count, Math.ceil(count / 2), spheres);
+
+    for (let z = 0; z < zn; ++z) {
+        for (let y = 0; y < yn; ++y) {
+            for (let x = 0; x < xn; ++x) {
```

Comment:
I wonder if we should add `Tensor.forEach` which would iterate based on actual axis ordering of the underlying data (slowest to fastest changing axis). This would prevent a lot of cache misses, although the drawback would be passing a callback for the inner body, which might erase the gains...

Reply:
made  #1426